ext/page_log/palm/palm_kv.c

```diff
@@ -0,0 +1,595 @@
+
+    memset(&kval, 0, sizeof(kval));
+    memset(&vval, 0, sizeof(vval));
+    memset(matches, 0, sizeof(*matches));
+    memset(&page_key, 0, sizeof(page_key));
+    memset(&result_key, 0, sizeof(result_key));
+    readonly_result_key = NULL;
+    now = palm_kv_timestamp_us();
+
+    matches->context = context;
+    matches->table_id = table_id;
+    matches->page_id = page_id;
+    matches->query_lsn = lsn;
+    matches->query_checkpoint_id = checkpoint_id;
+
+    page_key.table_id = table_id;
+    page_key.page_id = page_id;
+    page_key.lsn = lsn != 0 ? lsn : UINT64_MAX;
+    page_key.checkpoint_id = UINT64_MAX;
+    swap_page_key(&page_key, &page_key);
+    kval.mv_size = sizeof(page_key);
+    kval.mv_data = &page_key;
+    if ((ret = mdb_cursor_open(
+           context->lmdb_txn, context->env->lmdb_pages_dbi, &matches->lmdb_cursor)) != 0)
+        return (ret);
+    ret = mdb_cursor_get(matches->lmdb_cursor, &kval, &vval, MDB_SET_RANGE);
+    if (ret == MDB_NOTFOUND) {
+        /* If we went off the end, go to the last record. */
+        ret = mdb_cursor_get(matches->lmdb_cursor, &kval, &vval, MDB_LAST);
+    }
+    if (ret == 0) {
+        if (kval.mv_size != sizeof(PAGE_KEY))
+            return (EIO); /* not expected, data damaged, could be assert */
+        readonly_result_key = (PAGE_KEY *)kval.mv_data;
+        swap_page_key(readonly_result_key, &result_key);
+    }
+    /*
+     * Now back up until we get a match. This will be the last valid record that matches the
+     * table/page.
+     */
+    while (
+      ret == 0 && !RESULT_MATCH(&result_key, context, table_id, page_id, lsn, checkpoint_id, now)) {
+        ret = mdb_cursor_get(matches->lmdb_cursor, &kval, &vval, MDB_PREV);
+        readonly_result_key = (PAGE_KEY *)kval.mv_data;
+        swap_page_key(readonly_result_key, &result_key);
+    }
+    /*
+     * Now back up until we find the most recent full page that does not have a checkpoint more
+     * recent than asked for.
+     */
+    while (
+      ret == 0 && RESULT_MATCH(&result_key, context, table_id, page_id, lsn, checkpoint_id, now)) {
+        /* If this is what we're looking for, we're done, and the cursor is positioned. */
+        if (result_key.is_delta == false) {
+            matches->lsn = result_key.lsn;
+            matches->checkpoint_id = result_key.checkpoint_id;
+            matches->size = vval.mv_size;
+            matches->data = vval.mv_data;
+            matches->backlink_lsn = result_key.backlink_lsn;
+            matches->base_lsn = result_key.base_lsn;
+            matches->backlink_checkpoint_id = result_key.backlink_checkpoint_id;
+            matches->base_checkpoint_id = result_key.base_checkpoint_id;
+            matches->first = true;
+            return (0);
+        }
+        ret = mdb_cursor_get(matches->lmdb_cursor, &kval, &vval, MDB_PREV);
+        readonly_result_key = (PAGE_KEY *)kval.mv_data;
+        swap_page_key(readonly_result_key, &result_key);
+    }
```

Comment:
Why do we need to go to the first unmatched entry and then scroll back to the first entry? Can we just find first entry that suits our needs and return it?

Reply:
No, we need the last entry that has a checkpoint under a certain number. The checkpoint is only part of the key so I don't think we can just directly find a suitable entry.